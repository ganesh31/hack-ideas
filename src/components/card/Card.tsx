import { ThumbUpIcon } from "@heroicons/react/outline";
import invariant from "tiny-invariant";
import { Hack } from "../../types/hack";
import { Tag } from "../../types/tag";

interface Props {
  hack: Hack;
  onToggleLike: (hackId: number) => void;
  allTags: Tag[];
}

/**
 *
 * @param date time in milliseconds
 * @returns formatted date
 */
function dateToString(date: number): string {
  return new Date(date).toDateString();
}

function Card({ hack, onToggleLike, allTags }: Props) {
  return (
    <div className="group border-slate-800 hover:bg-slate-200 bg-slate-50 text-slate-800 hover:cursor-pointer min-w-[350px] sm:min-w-sm p-4 space-y-1 border rounded shadow">
      <div className="flex justify-between">
        <div className="text-lg font-bold">{hack.title}</div>
        <div className="flex space-x-1">
          <span data-testid="likes-count">{hack.likes || ""}</span>
          <ThumbUpIcon
            data-testid={"button-like"}
            onClick={() => onToggleLike(hack.id)}
            className={`${hack.likes !== 0 ? "text-red-600" : ""} w-5 h-5`}
          />
        </div>
      </div>
      <p className="text-ellipsis whitespace-nowrap overflow-hidden">
        {hack.description}
      </p>
      <div className="group-hover:opacity-100 flex justify-between text-sm transition duration-300 ease-in-out opacity-0">
        <span>
          <i>{dateToString(hack.createdAt)}</i>
        </span>
      </div>
      <div className="space-x-1 overflow-hidden">
        {allTags.length !== 0 &&
          hack.tags.map((id) => {
            const tag = allTags.find((tag) => tag.id === id);
            invariant(tag, "unknown tag");
            return <TagItem name={tag.name} key={id} />;
          })}
      </div>
    </div>
  );
}
function TagItem({ name }: { name: string }) {
  return (
    <span className="bg-slate-200 group-hover:bg-slate-50 whitespace-nowrap px-4 py-1 text-xs text-gray-600 rounded-md">
      {name}
    </span>
  );
}
export default Card;
