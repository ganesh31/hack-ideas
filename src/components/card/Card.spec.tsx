import { fireEvent, render, screen } from "@testing-library/react";
import Card from "./Card";

const allTags = [
  {
    id: 1,
    name: "recursion",
  },
  {
    id: 2,
    name: "math fundamentals",
  },
  {
    id: 3,
    name: "data structures",
  },
  {
    id: 4,
    name: "advanced",
  },
  {
    id: 5,
    name: "binary search",
  },
  {
    id: 6,
    name: "intermediate",
  },
];

const hack = {
  title: "Circa May '22",
  description: "Test your coding skills",
  authorId: 1,
  question: "### Question3",
  tags: [4, 3],
  createdAt: 1644825217436,
  id: 5,
  likes: 1,
  likedBy: [123],
};

describe("Card", () => {
  it("should call onToggle when user clicks on like button", () => {
    const mockToggle = jest.fn();

    const { asFragment } = render(
      <Card allTags={allTags} hack={hack} onToggleLike={mockToggle} />
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByTestId("button"));
    expect(mockToggle).toBeCalledTimes(1);
  });
});
