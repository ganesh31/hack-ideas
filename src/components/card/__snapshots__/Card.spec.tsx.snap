// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should call onToggle when user clicks on like button 1`] = `
<DocumentFragment>
  <div
    class="group border-slate-800 hover:bg-slate-200 bg-slate-50 text-slate-800 hover:cursor-pointer min-w-[350px] sm:min-w-sm p-4 space-y-1 border rounded shadow"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="text-lg font-bold"
      >
        Circa May '22
      </div>
      <div
        class="flex space-x-1"
      >
        <span
          data-testid="likes-count"
        >
          1
        </span>
        <svg
          aria-hidden="true"
          class="text-red-600 w-5 h-5"
          data-testid="button-like"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 10h4.764a2 2 0 011.789 2.894l-3.5 7A2 2 0 0115.263 21h-4.017c-.163 0-.326-.02-.485-.06L7 20m7-10V5a2 2 0 00-2-2h-.095c-.5 0-.905.405-.905.905 0 .714-.211 1.412-.608 2.006L7 11v9m7-10h-2M7 20H5a2 2 0 01-2-2v-6a2 2 0 012-2h2.5"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <p
      class="text-ellipsis whitespace-nowrap overflow-hidden"
    >
      Test your coding skills
    </p>
    <div
      class="group-hover:opacity-100 flex justify-between text-sm transition duration-300 ease-in-out opacity-0"
    >
      <span>
        <i>
          Mon Feb 14 2022
        </i>
      </span>
    </div>
    <div
      class="space-x-1 overflow-hidden"
    >
      <span
        class="bg-slate-200 group-hover:bg-slate-50 whitespace-nowrap px-4 py-1 text-xs text-gray-600 rounded-md"
      >
        advanced
      </span>
      <span
        class="bg-slate-200 group-hover:bg-slate-50 whitespace-nowrap px-4 py-1 text-xs text-gray-600 rounded-md"
      >
        data structures
      </span>
    </div>
  </div>
</DocumentFragment>
`;
